### PR TITLE
The Mighty Powered Drain

### PR DESCRIPTION
**Pull request by HenryLoenwind**
A new machine that picks up liquids from the world.

Should the recipe include a head?

Works best with it the core mod Water Hooks (https://github.com/HenryLoenwind/waterhook) installed, but doesn't require it.

Cannot be crafted on dedicated servers by default (config option),

### CODE REVIEW

File: src/main/java/crazypants/enderio/machine/drain/FluidHelper.java [hunk 1]
```diff
@@ -0,0 +1,328 @@
...
+      if (isSourceBlock(world, direction)) {
+        Fluid fluidForBlock = FluidRegistry.lookupFluidForBlock(direction.getBlock(world));
+        if (fluidForBlock != null) {
+          FluidHelper result = getInstance(world, new FluidStack(fluidForBlock, 1000), direction);
+          if (result != null) {
+            return result;
+          }
+        }
+      }
+    }
```
Comment (CrazyPants) [inline src/main/java/crazypants/enderio/machine/drain/FluidHelper.java]: Why are you running the loop twice?

File: src/main/java/crazypants/enderio/machine/drain/FluidHelper.java [hunk 2]
```diff
@@ -0,0 +1,328 @@
...
+      return false;
+    }
+  }
+
+  private static final ForgeDirection[] DIRECTIONS_INIT = {ForgeDirection.UP, ForgeDirection.NORTH, ForgeDirection.SOUTH, ForgeDirection.WEST, ForgeDirection.EAST};
+  
+  public static FluidHelper getInstance(World world, BlockCoord bc) {
+    for (ForgeDirection forgeDirection : DIRECTIONS_INIT) {
+      BlockCoord direction = bc.getLocation(forgeDirection);
+      if (isSourceBlock(world, direction)) {
```
Comment (CrazyPants) [inline src/main/java/crazypants/enderio/machine/drain/FluidHelper.java]: Why this check and not just != null, they both do the same thing anyway

**Comment by CrazyPants** [comment]
How well does it work without the core mod? Is it stable enough / performs well with out it?
Sound like a cool addition.

File: src/main/java/crazypants/enderio/machine/drain/FluidHelper.java (hunk 1 above)
Comment (HenryLoenwind) [inline src/main/java/crazypants/enderio/machine/drain/FluidHelper.java]: The logic is to first look for a source block, but if there's none to take a flowing block, too.

I could fold them into one loop like

    foreach (EPass pass : EPass.values()) {
        ....
        if (pass == EPass.PASS2 || isSourceBlock(world, direction)

but for just 2 passes and a couple of lines of code it looks like a big overkill. (And the reverse of the kind of loop unrolls any good optimizer does ;))

File: src/main/java/crazypants/enderio/machine/drain/FluidHelper.java (hunk 2 above)
Comment (HenryLoenwind) [inline src/main/java/crazypants/enderio/machine/drain/FluidHelper.java]: Sorry, I'm not 100% sure which of the lines that Github shows me you commented on.

The isSourceBlock() certainly is no simple null check; code is just above.

**Comment by HenryLoenwind** [comment]
Works fine here. I put it under some heavy stress testing (ocean draining) without issues. I tested it with vanilla and ender io fluids, and some modded ones I had in my dev env anyway.

Without the core mod, it will act as an infinite water source, no breakage. Just way more cpu usage than a Reservoir... ;)

What makes it heavy is vanilla's heavy liquid flowing code, or with Lava the heavy re-lighting.

**Comment by tterrag1098** [comment]
After some testing I found some annoyances in how it functions.

Mainly the "pulling" of sources towards it, while it is a bit neat, it is unnecessary lag for little aesthetic gain.

Finally, it doesn't seem to properly drain even "natural" lakes, which would be very annoying in gameplay.

**Comment by tterrag1098** [comment]
I am planning to do "the big merge" of the EnderCore branch soon, which means you will have to do a lot of work to sync up. If those are fixed soon I can pull this maybe.

**Comment by tterrag1098** [comment]
I am back from vacation, and nothing has been done on this, so I am putting it on hold so I can move forward with other things.

**Comment by HenryLoenwind** [comment]
STOP. Damaged branch, I'll fix later today!!!

**Comment by HenryLoenwind** [comment]
This is currently live in EnderIOAddons, if I bring this over here it'll be in a new branch.

**Pull request closed without merging**